### PR TITLE
use a finalizer instead of GC to clean up templateinstances

### PR DESCRIPTION
**Pull request by bparees**
bug https://bugzilla.redhat.com/show_bug.cgi?id=1568607

### CODE REVIEW

**Comment by bparees** [comment]
@deads2k @liggitt ptal

File: pkg/cmd/server/bootstrappolicy/controller_policy.go [hunk 1]
```diff
@@ -172,6 +172,19 @@ func init() {
 	addDefaultMetadata(&templateInstanceController)
 	controllerRoleBindings = append(controllerRoleBindings, templateInstanceController)
 
+	// template-instance-finalizer-controller
+	addControllerRole(rbac.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{Name: saRolePrefix + InfraTemplateInstanceFinalizerControllerServiceAccountName},
+		Rules: []rbac.PolicyRule{
+			rbac.NewRule("update").Groups(templateGroup).Resources("templateinstances/status").RuleOrDie(),
```
Comment (deads2k) [inline pkg/cmd/server/bootstrappolicy/controller_policy.go]: reading top to bottom, but I expected you to need broad delete powers.

File: pkg/template/apis/template/constants.go [hunk 2]
```diff
@@ -39,4 +39,12 @@ const (
...
 	// advertise the template as being bindable (default is true)
 	BindableAnnotation = "template.openshift.io/bindable"
+
+	// TemplateInstanceFinalizer is used to clean up the objects created by the template instance,
+	// when the template instance is deleted.
+	TemplateInstanceFinalizer = "template.openshift.io/finalizer"
+
+	// TemplateInstanceOwner is a label applied to all objects created from a template instance
+	// which contains the uid of the template instance.
+	TemplateInstanceOwner = "template.openshift.io/template-instance-owner"
```
Comment (deads2k) [inline pkg/template/apis/template/constants.go]: again reading top to bottom, but this screams: "migration problem" to me.

File: pkg/template/controller/templateinstance_controller.go [hunk 3]
```diff
@@ -72,18 +74,19 @@ type TemplateInstanceController struct {
...
-		config:           config,
-		kc:               kc,
-		templateClient:   templateClient,
-		buildClient:      buildClient,
-		lister:           informer.Lister(),
-		informer:         informer.Informer(),
-		queue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "openshift_template_instance_controller"),
-		readinessLimiter: workqueue.NewItemFastSlowRateLimiter(5*time.Second, 20*time.Second, 200),
-		clock:            clock.RealClock{},
+		restmapper:        restutil.DefaultMultiRESTMapper(),
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_controller.go]: this restmapper looks compiled in.  Why are you passing one in and using this?

File: pkg/template/controller/templateinstance_finalizer.go [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+	restmapper        meta.RESTMapper
+	dynamicRestMapper *discovery.DeferredDiscoveryRESTMapper
+	client            dynamic.DynamicInterface
+	config            *rest.Config
+	jsonConfig        *rest.Config
+	templateClient    templateclient.Interface
+
+	// FIXME: Remove then cient when the build configs are able to report the
+	//				status of the last build.
+	buildClient buildclient.Interface
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: top to bottom, but this looks weird.  Why?

File: pkg/template/controller/templateinstance_finalizer.go [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+
+	kc kclientsetinternal.Interface
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: top to bottom, but this looks weird.  Why?

File: pkg/template/controller/templateinstance_finalizer.go [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+	restutil "github.com/openshift/origin/pkg/util/rest"
+)
+
+// TemplateInstanceFinalizerController watches for new TemplateInstance objects and
+// instantiates the template contained within, using parameters read from a
+// linked Secret object.  The TemplateInstanceFinalizerController instantiates objects
+// using its own service account, first verifying that the requester also has
+// permissions to instantiate.
+type TemplateInstanceFinalizerController struct {
+	restmapper        meta.RESTMapper
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: why have this restmapper too?

File: pkg/template/controller/templateinstance_finalizer.go [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+	dynamicRestMapper *discovery.DeferredDiscoveryRESTMapper
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: why do you need to know it's a particular kind of restmapper?

File: pkg/template/controller/templateinstance_finalizer.go [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+	client            dynamic.DynamicInterface
+	config            *rest.Config
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: why are there two of these?

File: pkg/template/controller/templateinstance_finalizer.go [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+	templateClient    templateclient.Interface
+
+	// FIXME: Remove then cient when the build configs are able to report the
+	//				status of the last build.
+	buildClient buildclient.Interface
+
+	kc kclientsetinternal.Interface
+
+	lister   templatelister.TemplateInstanceLister
+	informer cache.SharedIndexInformer
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: this is unusual, why are you tracking it?

File: pkg/template/controller/templateinstance_finalizer.go [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+func (c *TemplateInstanceFinalizerController) sync(key string) error {
+	templateInstance, err := c.getTemplateInstance(key)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	needsFinalizing := false
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: check deletiontimestamp again.  If some cleared a finalizer, then manually deleted, then recreated with the same name, you'd end up here again and you should skip action.

File: pkg/template/controller/templateinstance_finalizer.go [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+			Kind:  o.Ref.Kind,
+		}
+		mapping, err := c.dynamicRestMapper.RESTMapping(gk, gv.Version)
+		if err != nil || mapping == nil {
+			errs = append(errs, fmt.Errorf("error mapping object %#v: %v", o, err))
+			continue
+		}
+
+		namespaced := mapping.Scope.Name() == meta.RESTScopeNameNamespace
+		apiResource := &metav1.APIResource{Name: mapping.Resource, Group: gv.Group, Version: gv.Version, Kind: o.Ref.Kind, Namespaced: namespaced}
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: why are you constructing this?

File: test/testdata/bootstrappolicy/bootstrap_cluster_role_bindings.yaml [hunk 12]
```diff
@@ -821,6 +821,36 @@ items:
...
+  - kind: ServiceAccount
+    name: template-instance-finalizer-controller
+    namespace: openshift-infra
+- apiVersion: rbac.authorization.k8s.io/v1beta1
+  kind: ClusterRoleBinding
+  metadata:
+    annotations:
+      rbac.authorization.kubernetes.io/autoupdate: "true"
+    creationTimestamp: null
+    name: admin
```
Comment (deads2k) [inline test/testdata/bootstrappolicy/bootstrap_cluster_role_bindings.yaml]: ah, I see.  Makes sense now.

Comment (deads2k) [reply test/testdata/bootstrappolicy/bootstrap_cluster_role_bindings.yaml]: > ah, I see. Makes sense now.

Name doesn't look good through.  `system:openshift:controller:<something>`

**Comment by deads2k** [comment]
I would encourage a unit test. I take it you're stuck on https://github.com/kubernetes/kubernetes/pull/63253 ?

File: pkg/cmd/server/bootstrappolicy/controller_policy.go (hunk 1 above)
Comment (bparees) [reply pkg/cmd/server/bootstrappolicy/controller_policy.go]: it's got project admin just like the template instance controller.  would you still expect something else?

File: pkg/template/apis/template/constants.go (hunk 2 above)
Comment (bparees) [reply pkg/template/apis/template/constants.go]: templateinstances created by old versions won't have this finalizer, and resources created by those templateinstances will have ownerrefs on them, which means they'll get handled by normal GC and the finalizer should ignore them.

templateinstances created by this version will have the finalizer and the objects created by the templateinstance will have the label and no ownerref, so GC will ignore them and finalization will cover it.

At least that is my theory.  Other than "i was creating a template instance but it's objects weren't created yet and then i upgraded my controller" do you see a hole?  I don't think i can convince myself to care about that scenario.

File: pkg/template/controller/templateinstance_controller.go (hunk 3 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_controller.go]: may have to talk in person about this one.  I didn't change this code, I only added the dynamicRestMapper.  (and per the discussion in the other PR, we might not need this mapper at all anymore but right now the checkReadiness codepath is using it)

File: pkg/template/controller/templateinstance_finalizer.go (hunk 6 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: cut+paste, will remove.

File: pkg/template/controller/templateinstance_finalizer.go (hunk 7 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: OCD :)

will change to interface.

File: pkg/template/controller/templateinstance_finalizer.go (hunk 8 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: cut+paste, will remove.

File: pkg/template/controller/templateinstance_finalizer.go (hunk 5 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: i must have pulled this from my sample finalizer and didn't use it here, will remove.

File: pkg/template/controller/templateinstance_finalizer.go (hunk 4 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: why does it look weird or why is it here? :)

looks like more cut+paste sloppiness on my part, will remove.

File: pkg/template/controller/templateinstance_finalizer.go (hunk 9 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: you mean vs adding my handler and throwing away the ref?  the controller Run waits for the cache to be synced.

Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: ie 	`if !cache.WaitForCacheSync(stopCh, c.informer.HasSynced) {`

File: pkg/template/controller/templateinstance_finalizer.go (hunk 11 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: dead code from an earlier attempt at this. will remove.

File: test/testdata/bootstrappolicy/bootstrap_cluster_role_bindings.yaml (hunk 12 above)
Comment (bparees) [reply test/testdata/bootstrappolicy/bootstrap_cluster_role_bindings.yaml]: i used the same name that templateinstancecontroller was using...should that be changed too?  (will reconciliation handle it?)

Comment (bparees) [reply test/testdata/bootstrappolicy/bootstrap_cluster_role_bindings.yaml]: (actually this is what `NewClusterBinding` gives back by default as the name..uses the rolename)

**Comment by bparees** [comment]
I wasn't aware of being stuck on anything other than your review comments.

I've hopefully addressed all your comments either in response form or in the new commit ("david 1").

Unit test for what aspect?

**Comment by bparees** [comment]
(and of course this PR sits on top of the CRD PR, so i need to get you happy w/ that PR before this one can merge)

File: pkg/template/controller/templateinstance_finalizer.go (hunk 9 above)
Comment (deads2k) [reply pkg/template/controller/templateinstance_finalizer.go]: We only keep a ref to the synced functions

Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: updated in `david 2`

File: test/integration/authorization_test.go [hunk 13]
```diff
@@ -622,7 +622,7 @@ func TestAuthorizationResourceAccessReview(t *testing.T) {
 				Users:           sets.NewString("edgar"),
 				Groups:          sets.NewString(),
 				Namespace:       "mallet-project",
-				EvaluationError: `[clusterrole.rbac.authorization.k8s.io "admin" not found, clusterrole.rbac.authorization.k8s.io "admin" not found]`,
+				EvaluationError: `[clusterrole.rbac.authorization.k8s.io "admin" not found, clusterrole.rbac.authorization.k8s.io "admin" not found, clusterrole.rbac.authorization.k8s.io "admin" not found]`,
```
Comment (bparees) [inline test/integration/authorization_test.go]: @enj i don't really understand why my changes in this PR would necessitate this change... can you enlighten me on what is going on here? (in general it seems weird that we'd return a string containing the same error message repeated N times w/ no additional context)

Comment (liggitt) [reply test/integration/authorization_test.go]: checking who can do a thing involves checking all applicable role bindings and the roles they refer to. any bindings that reference an unresolveable role result in a lookup error. all the lookup errors are aggregated here: https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac/subject_locator.go#L79

I'm not sure if we should dedupe the aggregated messages or add more info to distinguish which binding each lookup error came from (which could leak info).

In any case, you added a new clusterrolebinding which in this test becomes invalid (since the test intentionally deletes the admin role). This test might be better served creating a single-use role and a couple referencing rolebindings.

Comment (bparees) [reply test/integration/authorization_test.go]: thanks @liggitt, makes sense

File: pkg/template/apis/template/constants.go (hunk 2 above)
Comment (deads2k) [reply pkg/template/apis/template/constants.go]: I thought this was solving a cross namespace problem for you.  Don't you need to have the cluster update to be in a state where it's solved?

File: pkg/template/controller/templateinstance_finalizer.go [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+}
+
+// NewTemplateInstanceFinalizerController returns a new TemplateInstanceFinalizerController.
+func NewTemplateInstanceFinalizerController(dynamicRestMapper *discovery.DeferredDiscoveryRESTMapper, config *rest.Config, templateClient templateclient.Interface, informer internalversion.TemplateInstanceInformer) *TemplateInstanceFinalizerController {
+	c := &TemplateInstanceFinalizerController{
+		dynamicRestMapper: dynamicRestMapper,
+		templateClient:    templateClient,
+		lister:            informer.Lister(),
+		informerSynced:    informer.Informer().HasSynced,
+		queue:             workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "openshift_template_instance_controller"),
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: name

File: pkg/template/controller/templateinstance_finalizer.go [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+		UpdateFunc: func(_, obj interface{}) {
+			t := obj.(*templateapi.TemplateInstance)
+			if t.DeletionTimestamp != nil {
+				c.enqueue(t)
+			}
+		},
+	})
+
+	c.jsonConfig = rest.CopyConfig(config)
+	c.jsonConfig.ContentConfig = dynamic.ContentConfig()
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: don't need to set this

File: pkg/template/controller/templateinstance_finalizer.go [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+		},
+		UpdateFunc: func(_, obj interface{}) {
+			t := obj.(*templateapi.TemplateInstance)
+			if t.DeletionTimestamp != nil {
+				c.enqueue(t)
+			}
+		},
+	})
+
+	c.jsonConfig = rest.CopyConfig(config)
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: don't need to copy this

File: pkg/template/controller/templateinstance_finalizer.go [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	needsFinalizing := false
+	for _, v := range templateInstance.Finalizers {
+		if v == templateapi.TemplateInstanceFinalizer && templateInstance.DeletionTimestamp != nil {
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: move the deletiontimestamp check out of the for loop.  Just check above and return.

File: pkg/template/controller/templateinstance_finalizer.go [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+
+	glog.V(4).Infof("TemplateInstanceFinalizer controller: syncing %s", key)
+
+	errs := []error{}
+	background := metav1.DeletePropagationBackground
+	deleteOpts := &metav1.DeleteOptions{PropagationPolicy: &background}
+	for _, o := range templateInstance.Status.Objects {
+		glog.V(5).Infof("attempting to delete object: %#v", o)
+
+		gv, err := schema.ParseGroupVersion(o.Ref.APIVersion)
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: group and version should be different fields in your API.  API already baked?

File: pkg/template/controller/templateinstance_finalizer.go [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+		}
+
+		gvr := schema.GroupVersionResource{
+			Group:    gv.Group,
+			Version:  gv.Version,
+			Resource: mapping.Resource,
+		}
+
+		var resourceClient dynamic.DynamicResourceInterface
+		if namespaced {
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: this gets less ugly after the rebase.

File: pkg/template/controller/templateinstance_finalizer.go [hunk 20]
```diff
@@ -0,0 +1,261 @@
...
+	}
+	if !needsFinalizing {
+		return nil
+	}
+
+	glog.V(4).Infof("TemplateInstanceFinalizer controller: syncing %s", key)
+
+	errs := []error{}
+	background := metav1.DeletePropagationBackground
+	deleteOpts := &metav1.DeleteOptions{PropagationPolicy: &background}
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: why background?

File: pkg/template/controller/templateinstance_finalizer.go [hunk 21]
```diff
@@ -0,0 +1,261 @@
...
+
+	return nil
+}
+
+// Run runs the controller until stopCh is closed, with as many workers as
+// specified.
+func (c *TemplateInstanceFinalizerController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: glog.Info for starting, ending, and cache sync please.  They pop out once each and are super useful.

File: pkg/template/apis/template/constants.go (hunk 2 above)
Comment (bparees) [reply pkg/template/apis/template/constants.go]: not really, no.  anyone who created cross-namespace content already (which i suspect is no one) has already potentially experienced that content getting spontaneously erased by the GC (since as far as the GC is concerned, that content has no valid ownerref already).

We might do a knowledge base about cleaning up that potential content, but again i think cross-namespace object creation is a really rare use case.  In particular because it would have to be coupled with either TSB based instantiation, or direct templateinstance creation, which no one knows how to do except us.

File: pkg/template/controller/templateinstance_finalizer.go (hunk 19 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: yeah i saw, i'll clean it up at that point i guess.

File: pkg/template/controller/templateinstance_finalizer.go (hunk 20 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: didn't see a reason to hold up finalization waiting for GC to do its job on any child resources, but if that's not the right approach i'm happy to change it.

File: pkg/template/controller/templateinstance_finalizer.go [hunk 22]
```diff
@@ -0,0 +1,261 @@
...
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", templateInstance, err))
+		return
+	}
+
+	c.queue.Add(key)
+}
+
+// enqueueAfter adds a TemplateInstance to c.queue after a duration.
+func (c *TemplateInstanceFinalizerController) enqueueAfter(templateInstance *templateapi.TemplateInstance, duration time.Duration) {
```
[the remark this answers is not in the document]
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: was unused, removing.

File: pkg/template/controller/templateinstance_finalizer.go (hunk 14 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: fixing

File: pkg/template/controller/templateinstance_finalizer.go (hunk 15 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: removing

File: pkg/template/controller/templateinstance_finalizer.go (hunk 16 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: removing

File: pkg/template/controller/templateinstance_finalizer.go (hunk 17 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: fixing

File: pkg/template/controller/templateinstance_finalizer.go (hunk 21 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: adding

**Comment by bparees** [comment]
comments addressed or fixed in `david 4`

**Comment by bparees** [comment]
@deads2k putting this back on your radar

File: pkg/template/controller/templateinstance_finalizer.go (hunk 20 above)
Comment (deads2k) [reply pkg/template/controller/templateinstance_finalizer.go]: > didn't see a reason to hold up finalization waiting for GC to do its job on any child resources, but if that's not the right approach i'm happy to change it.

I think you want foreground.  It still sets a finalizer and you get an early return, but the object stays around until the descendants are removed.

File: pkg/template/controller/templateinstance_finalizer.go (hunk 18 above)
Comment (deads2k) [reply pkg/template/controller/templateinstance_finalizer.go]: > also i don't think we have a choice unless we move away from using an ObjectReference ?

Yeah, I'd avoid an object reference . We've defined specific types in most new code.

File: pkg/template/controller/templateinstance_finalizer.go [hunk 23]
```diff
@@ -0,0 +1,255 @@
...
+		gv, err := schema.ParseGroupVersion(o.Ref.APIVersion)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error parsing group version %s for object %#v: %v", o.Ref.APIVersion, o, err))
+			continue
+		}
+		gk := schema.GroupKind{
+			Group: gv.Group,
+			Kind:  o.Ref.Kind,
+		}
+		mapping, err := c.dynamicRestMapper.RESTMapping(gk, gv.Version)
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: make a comment that doing this means that if an api resource type is removed, this finalizer will fail here and hang.

Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: ok

File: pkg/template/controller/templateinstance_finalizer.go (hunk 18 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: yeah so that ship sailed when the templateinstance api was introduced.

**Comment by bparees** [comment]
comments addressed in `david 4`

File: pkg/template/controller/templateinstance_finalizer.go [hunk 24]
```diff
@@ -0,0 +1,259 @@
...
+		if err != nil || mapping == nil {
+			errs = append(errs, fmt.Errorf("error mapping object %#v: %v", o, err))
+			continue
+		}
+
+		namespaced := mapping.Scope.Name() == meta.RESTScopeNameNamespace
+		namespace := ""
+		if namespaced {
+			namespace = o.Ref.Namespace
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: just to confirm.  A normal user cannot touch this ref, right?

File: pkg/template/controller/templateinstance_finalizer.go [hunk 25]
```diff
@@ -0,0 +1,259 @@
...
+			resourceClient = c.client.ClusterResource(gvr)
+		}
+		err = resourceClient.Delete(o.Ref.Name, deleteOpts)
+		if err != nil && !errors.IsNotFound(err) {
+			errs = append(errs, fmt.Errorf("error deleting object %#v with mapping %#v: %v", o, mapping, err))
+			continue
+		}
+	}
+	if len(errs) > 0 {
+		return kerrs.NewAggregate(errs)
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: @liggitt do we allow status to be updated on objects being finalized?

Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: @bparees I strongly recommend emitting an event either here or on the general error handling for debugging later.

File: pkg/template/controller/templateinstance_finalizer.go [hunk 26]
```diff
@@ -0,0 +1,259 @@
...
+func (c *TemplateInstanceFinalizerController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.V(2).Infof("TemplateInstanceFinalizer controller waiting for cache sync")
+	if !cache.WaitForCacheSync(stopCh, c.informerSynced) {
+		return
+	}
+
+	glog.V(2).Infof("Starting TemplateInstanceFinalizer controller")
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_finalizer.go]: I recommend putting this at the info level so you know that you've sync'd.

File: test/extended/testdata/templates/templateservicebroker_bind.yaml [hunk 27]
```diff
@@ -61,18 +61,6 @@ objects:
           to:
             kind: Service
             name: service
-      - apiVersion: v1
```
Comment (deads2k) [inline test/extended/testdata/templates/templateservicebroker_bind.yaml]: why the change?

File: pkg/template/controller/templateinstance_finalizer.go (hunk 24 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: it's part of the templateinstance status so they should not be able to, no.

File: test/testdata/bootstrappolicy/bootstrap_cluster_role_bindings.yaml [hunk 28]
```diff
@@ -812,7 +812,7 @@ items:
     annotations:
       rbac.authorization.kubernetes.io/autoupdate: "true"
     creationTimestamp: null
-    name: admin
+    name: template-instance-controller:admin
```
Comment (deads2k) [inline test/testdata/bootstrappolicy/bootstrap_cluster_role_bindings.yaml]: `system:openshift:controller:template-instance-controller:admin`  "don't muck with this: it's for openshift: for a controller: this controller: bound to that role"

File: pkg/template/controller/templateinstance_finalizer.go (hunk 26 above)
Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: will do

File: test/testdata/bootstrappolicy/bootstrap_cluster_role_bindings.yaml [hunk 29]
```diff
@@ -821,6 +821,36 @@ items:
...
+  - kind: ServiceAccount
+    name: template-instance-finalizer-controller
+    namespace: openshift-infra
+- apiVersion: rbac.authorization.k8s.io/v1beta1
+  kind: ClusterRoleBinding
+  metadata:
+    annotations:
+      rbac.authorization.kubernetes.io/autoupdate: "true"
+    creationTimestamp: null
+    name: template-instance-finalizer-controller:admin
```
Comment (deads2k) [inline test/testdata/bootstrappolicy/bootstrap_cluster_role_bindings.yaml]: system:openshift:controller:etc

**Comment by deads2k** [comment]
minor comments.  lgtm otherwise

File: pkg/template/controller/templateinstance_finalizer.go (hunk 25 above)
Comment (liggitt) [reply pkg/template/controller/templateinstance_finalizer.go]: afaik, all updates (both to spec and to status) are allowed

Comment (bparees) [reply pkg/template/controller/templateinstance_finalizer.go]: ok

File: test/extended/testdata/templates/templateservicebroker_bind.yaml (hunk 27 above)
Comment (bparees) [reply test/extended/testdata/templates/templateservicebroker_bind.yaml]: it's really reverting a change (which was made in the other templateinstance PR).  the legacy/grouped type handling is checked in templateinstance_objectkinds.yaml, so i wanted to put this back the way i found it for simplicity.

**Comment by bparees** [comment]
comments addressed in `david 5`

**Comment by deads2k** [comment]
lgtm, please squash before merge

After the rebase we'll have a fake dynamic client.  I recommend writing a unit test.

**Comment by bparees** [comment]
squashed.

looks like the switch to foreground may be messing w/ one of the tests, will investigate if it fails again.
